src/legacy/ui/public/agg_types/controls/string.tsx

```diff
@@ -31,7 +31,7 @@ function StringParamEditor({
   setValue,
   setTouched,
 }: AggParamEditorProps<string>) {
-  if (aggParam.disabled && aggParam.disabled(agg)) {
+  if (aggParam.shouldShow && !aggParam.shouldShow(agg)) {
     // reset model value
     setValue();
```

Comment:
We should remove `setValue()` here because of missing default values set

Reply:
By just checking it the way you linked, we would still store that setting in the saved object (just not write it out into the DSL), e.g. if you switch from a terms aggregation on a string field where you have include/exclude entered to a field type not supporting include/exclude the input boxes will be hidden, the DSL will be correct but saving this visualization will still have an include/exclude key in it, which I think is not a good solution. But I agree, that I don't like the control actually taking care about resetting that value in case it's hidden.

I am wondering if we should just pull that up into a more generic way, e.g. everytime you switch a field/agg we will run through all the params check if any of them will be hidden and if so remove the value from of `agg.params` for it. That would essentially do the same behavior here. Maybe we can still do it in React, but at least move it up to the generic `AggParamReactWrapper`. This one could then also check `shouldShow` and just render `null` if it don't want to be rendered. That way we prevent every individual control to need to check for that too. What do you think @maryia-lapata ?

Reply:
@timroes thanks for suggestion. Please take a look at the latest changes.

Reply:
@timroes in that case that I can't use `string` type for `other bucket` and `missing bucket` labels, because it stores custom labels in the saved object even if checkboxes are unchecked..